File: springboot-backend/src/main/java/com/wedspot/backend/services/implementation/AuthService.java
package com.wedspot.backend.services.implementation;

import com.wedspot.backend.Model.*;
import com.wedspot.backend.Model.Entity.User;
import com.wedspot.backend.exception.ResourceAlreadyExistsException;
import com.wedspot.backend.exception.ResourceNotFoundException;
import com.wedspot.backend.mappers.UserMapper;
import com.wedspot.backend.repository.IAuthRepository;
import com.wedspot.backend.services.IAuthService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AuthService implements IAuthService {

    private final IAuthRepository authRepository;
    private final UserMapper userMapper;

    @Override
    public APIResponse login(LoginRequest request) {
        // Logic for login (Demo for now)
        System.out.println("Login request: " + request.getEmail());
        Map<String, Object> response = new HashMap<>();
        response.put("accessToken", "demo-jwt-token-" + request.getEmail().split("@")[0]);
        response.put("role", "Admin"); // Logic to fetch from DB
        response.put("name", request.getEmail().split("@")[0]);

        APIResponse apiResponse = new APIResponse();
        apiResponse.setData(response);
        apiResponse.setMessage("Login successful");
        return apiResponse;
    }

    @Override
    public APIResponse register(RegisterRequest request) {
        Optional<User> existingUser = authRepository.findByEmail(request.getEmail());
        if (existingUser.isPresent()) {
            throw new ResourceAlreadyExistsException("Email already exists");
        }
        User user = userMapper.toEntity(request);
        authRepository.save(user);

        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("User registered successfully");
        return apiResponse;
    }

    @Override
    public APIResponse getUser(Long userId) {
        User user = authRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        UserDTO userDTO = userMapper.toDTO(user);
        
        APIResponse apiResponse = new APIResponse();
        apiResponse.setData(userDTO);
        apiResponse.setMessage("User fetched successfully");
        return apiResponse;
    }

    @Override
    public APIResponse getAllUsers() {
        List<User> users = authRepository.findAll();
        List<UserDTO> userDTOs = users.stream()
                .map(userMapper::toDTO)
                .collect(Collectors.toList());
        
        APIResponse apiResponse = new APIResponse();
        apiResponse.setData(userDTOs);
        apiResponse.setTotalElements(userDTOs.size());
        apiResponse.setMessage("Users fetched successfully");
        return apiResponse;
    }

    @Override
    public APIResponse forgotPassword(ForgotPasswordRequest request) {
        // Logic to generate OTP and send email (Demo for now)
        System.out.println("Forgot password request for: " + request.getEmail());
        
        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("OTP sent to your email successfully!");
        return apiResponse;
    }

    @Override
    public APIResponse resetPassword(ResetPasswordRequest request) {
        // Logic to update password in DB (Demo for now)
        System.out.println("Reset password for: " + request.getEmail());
        
        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("Password reset successfully!");
        return apiResponse;
    }

    @Override
    public APIResponse verifyOtp(VerifyOtpRequest request) {
        // Logic to verify OTP (Demo for now)
        System.out.println("Verifying OTP for: " + request.getEmail() + " - " + request.getOtp());
        
        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("OTP verified successfully!");
        return apiResponse;
    }

    @Override
    public APIResponse verifyToken(String token) {
        // Logic to verify JWT token (Demo for now)
        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("Token is valid");
        return apiResponse;
    }

    @Override
    public APIResponse logout(String token) {
        // Logic to invalidate token (Demo for now)
        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("Logout successful");
        return apiResponse;
    }
}

File: backend/Services/Implementations/AuthService.cs
using backend.Exceptions;
using backend.Models;
using backend.Models.Entities;
using backend.Repositories;

namespace backend.Services.Implementations
{
    public class AuthService : IAuthService
    {
        private readonly IUserRepository _userRepository;

        public AuthService(IUserRepository userRepository)
        {
            _userRepository = userRepository;
        }



        public async Task<APIResponse> Login(LoginRequest request)
        {
            var user = await _userRepository.GetByEmailAsync(request.Email);
            if (user == null)
                throw new ResourceNotFoundException("User not found with this email");

            if (!BCrypt.Net.BCrypt.Verify(request.Password, user.Password))
                throw new InvalidCredentialsException("Invalid password. Please try again.");

            return new APIResponse
            {
                Message = "Login successful",
                Data = new LoginResponse
                {
                    Id = user.Id,
                    AccessToken = $"demo-jwt-token-{user.Name}",
                    Role = user.Role,
                    Name = user.Name,
                    Email = user.Email
                }
            };
        }

        public async Task<APIResponse> Register(RegisterRequest request)
        {
            var existingUser = await _userRepository.GetByEmailAsync(request.Email);

            if (existingUser != null)
                throw new ResourceAlreadyExistsException("An account with this email already exists");

            var user = new User
            {
                Name = request.Name,
                Email = request.Email,
                Password = BCrypt.Net.BCrypt.HashPassword(request.Password),
                Role = request.Role,
                PhoneNumber = request.PhoneNumber,
                Location = request.Location
            };

            await _userRepository.AddAsync(user);

            return new APIResponse
            {
                Message = "User registered successfully",
                Data = new LoginResponse
                {
                    Id = user.Id,
                    AccessToken = $"demo-jwt-token-{user.Name}",
                    Role = user.Role,
                    Name = user.Name,
                    Email = user.Email
                }
            };
        }

        public async Task<APIResponse> ForgotPassword(ForgotPasswordRequest request)
        {
            var user = await _userRepository.GetByEmailAsync(request.Email);

            if (user == null)
                throw new ResourceNotFoundException("No account found with this email");

            return new APIResponse { Message = "OTP sent to your email" };
        }

        public async Task<APIResponse> ResetPassword(ResetPasswordRequest request)
        {
            var user = await _userRepository.GetByEmailAsync(request.Email);

            if (user == null)
                throw new ResourceNotFoundException("No account found with this email");

            user.Password = BCrypt.Net.BCrypt.HashPassword(request.Password);
            await _userRepository.UpdateAsync(user);

            return new APIResponse { Message = "Password reset successfully" };
        }

        public Task<APIResponse> VerifyOtp(VerifyOtpRequest request)
        {
            return Task.FromResult(new APIResponse { Message = "OTP verified successfully" });
        }

        public Task<APIResponse> VerifyToken(string token)
        {
            if (string.IsNullOrWhiteSpace(token))
                throw new InvalidCredentialsException("Token is missing or invalid");

            return Task.FromResult(new APIResponse { Message = "Token is valid" });
        }

        public Task<APIResponse> Logout(string token)
        {
            return Task.FromResult(new APIResponse { Message = "Logout successful" });
        }
    }
}
